### PR TITLE
refactor(deps): Pin all dependencies to the lowest possible versions

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/nextcloud/neon/issues/932

Local renovate run confirms this works.

### CODE REVIEW

File: packages/nextcloud/pubspec.yaml [hunk 1]
```diff
@@ -6,37 +6,37 @@ environment:
...
+  version: ^3.0.0
+  xml: ^6.0.0
+  xml_annotation: ^2.1.0
 
 dev_dependencies:
-  build_runner: ^2.4.6
-  built_value_generator: ^8.6.3
-  coverage: ^1.6.4
+  build_runner: ^2.0.0
+  built_value_generator: ^8.0.0
```
Comment (Leptopoda) [inline packages/nextcloud/pubspec.yaml:28]: We'll need at least 8.5 for the string interpolation issue I fixed upstream
https://pub.dev/packages/built_value_generator/changelog#850

**Review by Leptopoda — CHANGES_REQUESTED** [review]
Maybe we should keep the dev dependencies at their current versions.

They shouldn't affect devs using our packages.

**Comment by provokateurin** [comment]
Ah I forgot to run all the build_runner things, so maybe I need to raise some constraints again

**Comment by provokateurin** [comment]
I think we should run `melos exec flutter pub downgrade` before running any of the linting and tests. That way we can ensure that our constraints work fine and we are not running on bad versions.

**Comment by provokateurin** [comment]
> I think we should run melos exec flutter pub downgrade before running any of the linting and tests. That way we can ensure that our constraints work fine and we are not running on bad versions.

This doesn't work, because other packages have broken dependency constraints that break us. We can't do anything about that, so we need to be careful around this. Though now we have the minimums figured out once, we just need to only upgrade when needed.

File: docs/dependencies.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Dependencies
+
+We follow the guidelines outlined in https://dart.dev/tools/pub/dependencies with some rules and automation:
+1. We use the caret notation with the major version set to the latest major version and the minor and patch versions set to 0. In case we required a new feature or bug fix, updating the constraint to the minimum version that contains the particular feature or bug fix is needed.
+2. Dev dependencies should be pinned to the latest versions since they do not affect consumers of our packages.
+3. Using Renovate we automatically update our dependency constraints. For non-dev dependencies this will be the latest major version, for everything else it will upgrade to the latest minor and patch versions as well. The lockfiles are also kept up-to-date with Renovate to compile everything with the latest available versions.
+4. To validate that our version constraints are correct, we run our linting and tests against the lowest dependency versions that are allowed by our constraints.
```
[the remark this answers is not in the document]
Comment (provokateurin) [reply docs/dependencies.md:7]: Ah forgot to remove that again

**Review by Leptopoda — COMMENTED** [review]
I think we can still keep the dependencies of `App` pinned to the latest.
Similar to the root package as this won't be published.

In theory this doesn't change anything as we update the lockfile anyways but reading the pubspec.yaml is easier than the lockfile.

**Comment by provokateurin** [comment]
I'm not sure how easy it is to configure that in renovate.
I would argue we can drop the constraints completely, since we want to use the versions that are selected by the constraints of the other packages and we don't want to affect the constraint resolving :thinking:

**Comment by Leptopoda** [comment]
Sounds like a good idea; Let's try this.

**Review by Leptopoda — APPROVED** [review]

**Pull request merged**